src/PoolManager.sol

```diff
@@ -207,34 +218,71 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC6909Claims {
+
+        bool exactInput = params.amountSpecified > 0;
+        params.amountSpecified -= hookDeltaInSpecified;
+        require(exactInput ? params.amountSpecified > 0 : params.amountSpecified < 0);
+
+        // ExactIn (amountSpecified is positive)
+        // In the case where the hook contributed x, x is given to the user now. After the swap amountSpecified + x input are taken from the user, meaning the user pays amountSpecified overall.
+        // In the case where the hook took x, x is taken from the user now. After the swap amountSpecified - x input are taken from the user, meaning the user pays amountSpecified overall.
+        // ExactOut (amountSpecified is negative)
+        // In the case where the hook contributed x, x output is given to the user now. The swap of amountSpecified + x is swapping for x fewer output tokens - so the user gets amountSpecified overall.
+        // In the case where the hook took x, x output is taken from the user now. The swap of amountSpecified - x is swapping for x extra output tokens - so the user gets amountSpecified overall.
+        _accountDeltaFor(
+            (exactInput == params.zeroForOne) ? key.currency0 : key.currency1, hookDeltaInSpecified, address(key.hooks)
         );
 
-        _accountPoolBalanceDelta(key, delta);
-
-        // the fee is on the input currency
-        unchecked {
-            if (feeForProtocol > 0) {
-                protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+        {
+            uint256 feeForProtocol;
+            uint24 swapFee;
+            Pool.SwapState memory state;
+            (delta, feeForProtocol, swapFee, state) = pools[id].swap(
+                Pool.SwapParams({
+                    tickSpacing: key.tickSpacing,
+                    zeroForOne: params.zeroForOne,
+                    amountSpecified: params.amountSpecified,
+                    sqrtPriceLimitX96: params.sqrtPriceLimitX96
+                })
+            );
+
+            // the fee is on the input currency
+            unchecked {
+                if (feeForProtocol > 0) {
+                    protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+                }
             }
+
+            // TODO returning deltas here doesnt make any sense
+            emit Swap(
+                id,
+                msg.sender,
+                -delta.amount0(),
+                -delta.amount1(),
+                state.sqrtPriceX96,
+                state.liquidity,
+                state.tick,
+                swapFee
+            );
         }
 
-        emit Swap(
-            id, msg.sender, -delta.amount0(), -delta.amount1(), state.sqrtPriceX96, state.liquidity, state.tick, swapFee
+        (int128 hookDeltaInUnspecified) = key.hooks.afterSwap(key, params, delta, hookData);
+        _accountDeltaFor(
+            (exactInput == params.zeroForOne) ? key.currency1 : key.currency0,
```

Comment:
this check is done 3 times and honestly its a bit hard to read/understand.. took me a minute to verify exactly what was happening. maybe saving it as a var would help readability or writing a helper.. something like `getSpecifiedCurrency()`?

Or actually even better I think we could even add this to a library that handles the adding/subtracting of hook deltas and all the currencyIn/Out/Specified stuff and so eventually in the main path we could end up with something like:

```
(amount0, amount1) = delta.applyHookDelta();
_accountDeltaFor(currencyUnspecified, hookDeltaUnspecified, hook)
delta = delta + toBalanceDelta(amount0, amount1)
```
fwiw might also have to save an extra couple of vars like currencyUnspecified, currencySpecified but I think thats ok bc we could combine the multiple checks with (exactInput == params.zeroForOne) into one:

```
currencySpecified, currencyUnspecified = (exactInput == params.zeroForOne) ? (key.currency0, key.currency1) : (key.currency1 ,key.currency0)